lib/widgets/stream_colors.dart

```diff
@@ -0,0 +1,223 @@
+import 'dart:ui';
+
+import 'package:flutter/material.dart';
+import 'package:flutter_color_models/flutter_color_models.dart';
+
+import '../api/model/model.dart';
+import 'color.dart';
+
+/// A provider for [StreamColorSwatch]es that computes and caches them.
+abstract class StreamColorSwatches {
+  static final StreamColorSwatches light = _StreamColorSwatchesLight._();
+  static final StreamColorSwatches dark = _StreamColorSwatchesDark._();
+
+  final Map<int, StreamColorSwatch> _cache = {};
+
+  /// Gives the cached [StreamColorSwatch] for a [subscription.color].
+  StreamColorSwatch forBaseColor(int base) =>
+    _cache[base] ??= computeForBaseColor(base);
+
+  StreamColorSwatch computeForBaseColor(int base);
+
+  /// Gives a [StreamColorSwatches], lerped to [other] at [t].
+  ///
+  /// If [this] and [other] are [identical], returns [this].
+  ///
+  /// Else returns an instance whose [forBaseColor] will call
+  /// [this.forBaseColor] and [other.forBaseColor]
+  /// and return [StreamColorSwatch.lerp]'s result on those.
+  /// This computation is cached on the instance
+  /// in order to save work building [t]'s animation frame when there are
+  /// multiple UI elements using the same [subscription.color].
+  StreamColorSwatches lerp(StreamColorSwatches? other, double t) {
+    // This short-circuit helps when [a] and [b]
+    // are both [StreamColorSwatches.light]
+    // or both [StreamColorSwatches.dark].
+    // Empirically, [lerp] is called even when the theme hasn't changed,
+    // so this is an important optimization.
+    if (identical(this, other)) {
+      return this;
+    }
+
+    return _StreamColorSwatchesLerped(this, other, t);
+  }
+}
+
+/// The [StreamColorSwatches] for the light theme.
+class _StreamColorSwatchesLight extends StreamColorSwatches {
+  _StreamColorSwatchesLight._();
```

Comment:
nit:
```suggestion
  _StreamColorSwatchesLight();
```

Has the same effect since the class is already private, and makes the syntax a bit cleaner at the call site.